script/tool/lib/src/publish_command.dart

```diff
@@ -415,42 +419,46 @@ If running this command on CI, you can set the pub credential content in the $_p
       ..writeStringSync(credential)
       ..closeSync();
   }
-
-  /// Returns the correct path where the pub credential is stored.
-  @visibleForTesting
-  static String getCredentialPath() {
-    return _credentialsPath;
-  }
 }
 
 /// The path in which pub expects to find its credentials file.
-final String _credentialsPath = () {
-  // This follows the same logic as pub:
-  // https://github.com/dart-lang/pub/blob/d99b0d58f4059d7bb4ac4616fd3d54ec00a2b5d4/lib/src/system_cache.dart#L34-L43
-  String? cacheDir;
-  final String? pubCache = io.Platform.environment['PUB_CACHE'];
-  if (pubCache != null) {
-    cacheDir = pubCache;
-  } else if (io.Platform.isWindows) {
-    final String? appData = io.Platform.environment['APPDATA'];
+String _getCredentialsPath(
+    {required Platform platform, required p.Context path}) {
+  // See https://github.com/dart-lang/pub/blob/master/doc/cache_layout.md#layout
+  String? configDir;
+  if (platform.isLinux) {
+    String? configHome = platform.environment['XDG_CONFIG_HOME'];
+    if (configHome == null) {
+      final String? home = platform.environment['HOME'];
+      if (home == null) {
+        printError('"HOME" environment variable is not set.');
+      } else {
+        configHome = path.join(home, '.config');
+      }
+    }
+    if (configHome != null) {
+      configDir = path.join(configHome, 'dart');
+    }
+  } else if (platform.isWindows) {
+    final String? appData = platform.environment['APPDATA'];
     if (appData == null) {
       printError('"APPDATA" environment variable is not set.');
     } else {
-      cacheDir = p.join(appData, 'Pub', 'Cache');
+      configDir = path.join(appData, 'dart');
     }
-  } else {
-    final String? home = io.Platform.environment['HOME'];
+  } else if (platform.isMacOS) {
+    final String? home = platform.environment['HOME'];
     if (home == null) {
       printError('"HOME" environment variable is not set.');
     } else {
-      cacheDir = p.join(home, '.pub-cache');
+      configDir = path.join(home, 'Library', 'Application Support', 'dart');
     }
   }
 
-  if (cacheDir == null) {
-    printError('Unable to determine pub cache location');
+  if (configDir == null) {
+    printError('Unable to determine pub con location');
     throw ToolExit(1);
   }
 
-  return p.join(cacheDir, 'credentials.json');
-}();
+  return path.join(configDir, 'pub-credentials.json');
```

Comment:
nit: You've changed p -> path in multiple places, but the import is still `as p`.

Reply:
That's actually correct; `path` is a `p.Context` instance constructed based on the `platform` instance, which means it follows the mock platform if one is injected by tests.

What makes this extra confusing I think is that the `path` package provides everything as both methods on a context, and free functions (which call the method of the same name on the default context). So calling the free functions with a prefixed import and calling the methods on an instance look the same, even though they are actually calling different things.